test/unit/org/apache/cassandra/db/guardrails/CassandraPasswordValidatorTest.java

```diff
@@ -0,0 +1,245 @@
+                                        MIN_LOWER_CASE_CHARS_WARN_KEY, 2,
+                                        MIN_CHARACTERISTICS_WARN_KEY, 3,
+                                        MIN_CHARACTERISTICS_FAIL_KEY, 2,
+                                        MIN_LENGTH_WARN_KEY, 3,
+                                        MIN_LENGTH_FAIL_KEY, 2),
+                           format("The shortest password to pass the warning validator for any %s characteristics out of %s is %s but you have set the %s to %s.",
+                                  3, MAX_CHARACTERISTICS, 4, MIN_LENGTH_WARN_KEY, 3));
+
+        validateWithConfig(() ->
+                           new HashMap<>()
+                           {{
+                               put(MIN_SPECIAL_CHARS_FAIL_KEY, 1);
+                               put(MIN_DIGITS_CHARS_WARN_KEY, 2);
+                               put(MIN_DIGITS_CHARS_FAIL_KEY, 1);
+                               put(MIN_UPPER_CASE_CHARS_WARN_KEY, 2);
+                               put(MIN_UPPER_CASE_CHARS_FAIL_KEY, 1);
+                               put(MIN_LOWER_CASE_CHARS_WARN_KEY, 2);
+                               put(MIN_LOWER_CASE_CHARS_FAIL_KEY, 1);
+                               put(MIN_CHARACTERISTICS_WARN_KEY, 4);
+                               put(MIN_CHARACTERISTICS_FAIL_KEY, 3);
+                               put(MIN_LENGTH_WARN_KEY, 8);
+                               put(MIN_LENGTH_FAIL_KEY, 2);
+                           }},
+                           format("The shortest password to pass the failing validator for any %s characteristics out of %s is %s but you have set the %s to %s.",
+                                  3, MAX_CHARACTERISTICS, 3, MIN_LENGTH_FAIL_KEY, 2));
+    }
+
+
+    @Test
+    public void testIllegalSequences()
+    {
+        CassandraPasswordValidator validator = new CassandraPasswordValidator(new CustomGuardrailConfig());
+        Optional<ValidationViolation> validationResult = validator.shouldFail("A1$abcdefgh");
+        assertTrue(validationResult.isPresent());
+        assertThat(validationResult.get().message, containsString("Password contains the illegal alphabetical sequence 'abcdefgh'."));
+
+        validationResult = validator.shouldFail("A1$a123456");
+        assertTrue(validationResult.isPresent());
+        assertThat(validationResult.get().message, containsString("Password contains the illegal numerical sequence '123456'."));
+
+        validationResult = validator.shouldFail("A1$asdfghjkl");
+        assertTrue(validationResult.isPresent());
+        assertThat(validationResult.get().message, containsString("Password contains the illegal QWERTY sequence 'asdfghjkl'."));
+    }
+
+    @Test
+    public void testWhitespace()
+    {
+        CassandraPasswordValidator validator = new CassandraPasswordValidator(new CustomGuardrailConfig());
+        Optional<ValidationViolation> validationResult = validator.shouldFail("A1$abcd efgh");
```

Comment:
should we allow spaces? I guess a more general question is whether we should allow configuring the special characters via cassandra.yaml

Reply:
I do not think we should allow spaces. Let's just make it simple. Passay's whitespace rule rejects this:

    protected static final char[] CHARS = new char[]{'\t', '\n', '\u000b', '\f', '\r', ' '};

I don't know who "sane" would want to use these characters in a password. It means that we would enable them to be generated as well. Then a user would have quite a hard time to figure out how to enter it upon login. Not talking about how they would actually save it to e.g. cqlshrc and how cqlsh would interpret that. That's a can of worms I am not super keen on opening.

For special characters, we might enable it to be configurable with the defaults as provided. One caveat thought. For simplicity, I would not support illegal character to be " ' " (single quote). Because if you go to do this:

    create role 'stefan' with password = 'abcd'efgh'

notice that we would support `'` as valid character, but CQLSH would think that your password is `abcd` and it will reject everything after `'` and it will fail the query.

Reply:
one of the passwords generated for me during testing had a backtick `` ` ``  and I had a hard time logging in with that password, so I see your point.